plutus-core/plutus-core/src/PlutusCore/Check/Normal.hs

```diff
@@ -43,20 +48,35 @@ check Var{}                  = pure ()
 check Constant{}             = pure ()
 check Builtin{}              = pure ()
 
-isNormalType :: Type tyname uni ann -> Bool
+isNormalType :: HasUniApply uni => Type tyname uni ann -> Bool
 isNormalType = isRight . normalType
 
-normalType :: Type tyname uni ann -> Either (NormCheckError tyname name uni fun ann) ()
+normalType
+    :: HasUniApply uni
+    => Type tyname uni ann -> Either (NormCheckError tyname name uni fun ann) ()
 normalType (TyFun _ i o)       = normalType i >> normalType o
 normalType (TyForall _ _ _ ty) = normalType ty
 normalType (TyIFix _ pat arg)  = normalType pat >> normalType arg
 normalType (TySOP _ tyls)      = traverse_ (traverse_ normalType) tyls
 normalType (TyLam _ _ _ ty)    = normalType ty
--- See Note [PLC types and universes].
-normalType TyBuiltin{}         = pure ()
 normalType ty                  = neutralType ty
 
-neutralType :: Type tyname uni ann -> Either (NormCheckError tyname name uni fun ann) ()
-neutralType TyVar{}           = pure ()
-neutralType (TyApp _ ty1 ty2) = neutralType ty1 >> normalType ty2
-neutralType ty                = Left (BadType (typeAnn ty) ty "neutral type")
+neutralType
+    :: HasUniApply uni
+    => Type tyname uni ann -> Either (NormCheckError tyname name uni fun ann) ()
+neutralType TyVar{}                 = pure ()
+neutralType (TyBuiltin ann someUni) = neutralUni ann someUni
+neutralType (TyApp _ ty1 ty2)       = neutralType ty1 >> normalType ty2
+neutralType ty                      = Left (BadType (typeAnn ty) ty "neutral type")
+
+-- See Note [Normalization of built-in types].
```

Comment:
The Note:

```
{- Note [Normalization of built-in types]
Instantiating a polymorphic built-in type amounts to applying it to some arguments. However,
the notion of "applying" is ambiguous, it can mean one of these two things:

1. lifting the built-in type to 'Type' and applying that via 'TyApp'
2. applying the built-in type right inside the universe to get a monomorphized type tag
   (e.g. the default universe has 'DefaultUniApply' for that purpose)

We need both of these things. The former allows us to assign types to polymorphic built-in functions
(otherwise applying a built-in type to a type variable would be unrepresentable), the latter is
used at runtime to juggle type tags so that we can avoid @unsafeCoerce@-ing, bring instances in
scope via 'bring' etc -- for all of that we have to have fully monomorphized type tags at runtime.

So in order for type checking to work we need to normalize polymorphic built-in types. For that
we simply turn intra-universe applications into regular type applications during type normalization.

We could go the other way around and "reduce" regular type applications into intra-universe ones,
however that would be harder to implement, because collapsing a general 'Type' into a 'SomeTypeIn'
is harder than expanding a 'SomeTypeIn' into a 'Type'. And it would also be impossible to do in the
general case, 'cause you can't collapse an application of a built-in type to, say, a type variable
into an intra-universe application as there are no type variables there. I guess we could "reduce"
application of built-in types in some cases and not reduce them in others and still make the whole
thing work, but that requires substantially more logic and is also a lot harder to get right.
Hence we do the opposite, which is straightforward.
-}
```